_adr/83/index.adoc

```diff
@@ -0,0 +1,155 @@
+The only way to do that, is to perform the same checks that Kas Fleet Manager will do while executing the action.
+
+This brings complexities and duplications. For example, just to check whether the creation of a new kafka is possible or not,
+currently UI and CLI have to perform at least the following checks:
+
+. Verify that the required terms and conditions have been accepted (this requires interaction with another service)
+. Make sure you have enough quota allocated. This requires:
+.. Having knowledge of the AMS endpoints and how to deal with them
+.. Having knowledge of the AMS terms used for designing the instance types (RHOSAKTrial for eval/developer and RHOSAK for standard)
+.. Enquire the quota assigned using the AMS endpoints, loop through the resources and check that enough quota is still available for product RHOSAK and resource RHOSAK
+.. If you have no quota assigned for STANDARD instances (RHOSAK), fallback to EVAL
+.. If you have quota for STANDARD (RHOSAK), but it is 0, fallback to EVAL (RHOSAKTrial)
+.. If you have quota for STANDARD (RHOSAK) and it is greater than 0 but all the granted quota is consumed, the creation should fail
+. Verify that at least one OCD Cluster has capacity for the instance you are going to create
+
+Another point of attention is that currently, any change in the flow requires an update to all the actors/consumers who are carrying out those checks.
+
+## Goals
+* API consumers can query a new KAS Fleet Manager Permission API to check if they can perform a certain action and receive a response with the outcome (allowed/denied) containing all the granted resources or, eventually, the missing requirements.
+
+## Non-goals
+
+## Current architecture
+* Kas Fleet Manager currently only allows the ‘try and see if it works’ approach: if what you are trying to do fails, Kas Fleet Manager gives you some details on went wrong, but you have no way of checking in advance if what you want to do is possible
+* API Consumers (like UI) are doing the required checkss into their codebase and every time a change occurs, both Kas Fleet Manager and UI have to be updated to reflect the change
+
+## Proposed architecture
+* We need to create a new API.
+The API will be authenticated and will allow the caller to check in advance whether he can perform a certain operation.
+A different endpoint will be provided for each verb-resource pair, wich, in turn, will provide a schema defining exactly what the answer will look like.
++
+```
+GET /api/v1/can/{verb}/{resource}?deep=true
+```
++
+The API will stop at the first error unless `deep=true` is specified
+Output of the API will be a JSON with at least the following content:
++
+```
+{
+  allowed: boolean,
+  reasons: [{
+    code: string,
+    description: string,
+    metadata: {             // additional details
+      {key string}: {value string}
+    }
+  }]
+}
+```
```

Comment:
Is this JSON out for when deep=true? If not, is it possible to have an example of when deep=true?

Reply:
This is the output both when deep=true and deep=false.
The difference is that when deep=true, the `reasons` array will always have a size of 1

Reply:
Thanks Massimo. Is it possible for you to share an example of both when deep=false and deep=true to make sure I really get the difference? Thanks

Reply:
here: https://github.com/bf2fc6cc711aee1a0c2a/architecture/pull/52/files#diff-01aa0bf716b701e03b5bacdf09f7b0deab40d658c388fa3e272db1b6255f5cd5R115

Reply:
Perfect, thanks. I somehow missed that. 

So when `deep=false`, the same example might look like
```
{
  allowed: false,
  reasons: [{
    code: "ERR-TERMS-01",
    description: "Required terms have not been accepted",
    metadata: {
      acceptance_url: "https://url.to.accept.terms"
    }}
  ]
}
```
because we stopped at the first error which was encountered, is my understanding correct?

Reply:
Exactly

Reply:
Thank you.